core/src/com/unciv/logic/city/CityConstructions.kt

```diff
@@ -479,16 +480,72 @@ class CityConstructions : IsPartOfGameInfoSerialization {
     }
 
     fun addBuilding(buildingName: String) {
-        val buildingObject = city.getRuleset().buildings[buildingName]!!
-        builtBuildingObjects = builtBuildingObjects.withItem(buildingObject)
+        val building = city.getRuleset().buildings[buildingName]!!
+        val civ = city.civ
+
+        if (building.cityHealth > 0) {
+            // city built a building that increases health so add a portion of this added health that is
+            // proportional to the city's current health
+            city.health += (building.cityHealth.toFloat() * city.health.toFloat() / city.getMaxHealth().toFloat()).toInt()
+        }
+        builtBuildingObjects = builtBuildingObjects.withItem(building)
         builtBuildings.add(buildingName)
+
+        /** Support for [UniqueType.CreatesOneImprovement] */
+        applyCreateOneImprovement(building)
+
+        addFreeBuildings()
+
+        val triggerNotificationText ="due to constructing [$buildingName]"
+
+        for (unique in building.uniqueObjects)
+            if (!unique.hasTriggerConditional())
+                UniqueTriggerActivation.triggerCivwideUnique(unique, civ, city, triggerNotificationText = triggerNotificationText)
+
+
+        for (unique in civ.getTriggeredUniques(UniqueType.TriggerUponConstructingBuilding, StateForConditionals(civ, city)))
+            if (unique.conditionals.any {it.type == UniqueType.TriggerUponConstructingBuilding && building.matchesFilter(it.params[0])})
+                UniqueTriggerActivation.triggerCivwideUnique(unique, civ, city, triggerNotificationText = triggerNotificationText)
+
+        for (unique in civ.getTriggeredUniques(UniqueType.TriggerUponConstructingBuildingCityFilter, StateForConditionals(civ, city)))
+            if (unique.conditionals.any {it.type == UniqueType.TriggerUponConstructingBuildingCityFilter
+                    && building.matchesFilter(it.params[0])
+                    && city.matchesFilter(it.params[1])})
+                UniqueTriggerActivation.triggerCivwideUnique(unique, civ, city, triggerNotificationText = triggerNotificationText)
```

Comment:
This is starting to be a big function, I feel that this part of it is significant enough to extract to a separate function (triggerNotificationsForNewBuilding), YMMV